// @excalidraw/mermaid-to-excalidraw/dist/parser/state.js
import { CONTAINER_STYLE_PROPERTY, LABEL_STYLE_PROPERTY, } from "../interfaces.js";
import { computeEdgePositions, entityCodesToText, getTransformAttr, } from "../utils.js";
import { cleanCSSValue, isValidCSSColor, parseCSSDeclarations, } from "./cssUtils.js";
const isMeaningfulCSSColor = (value) => {
    const normalizedValue = cleanCSSValue(value || "");
    if (!normalizedValue) {
        return false;
    }
    if (normalizedValue === "none" ||
        normalizedValue === "transparent" ||
        normalizedValue === "rgba(0, 0, 0, 0)" ||
        normalizedValue === "rgba(0,0,0,0)") {
        return false;
    }
    return isValidCSSColor(normalizedValue);
};
const applyContainerStyleProperty = (style, property, value) => {
    switch (property) {
        case CONTAINER_STYLE_PROPERTY.FILL:
        case CONTAINER_STYLE_PROPERTY.STROKE:
            if (isMeaningfulCSSColor(value)) {
                style[property] = cleanCSSValue(value);
            }
            break;
        case CONTAINER_STYLE_PROPERTY.STROKE_WIDTH:
        case CONTAINER_STYLE_PROPERTY.STROKE_DASHARRAY:
            if (cleanCSSValue(value)) {
                style[property] = cleanCSSValue(value);
            }
            break;
    }
};
const applyLabelStyleProperty = (style, property, value) => {
    if (property === LABEL_STYLE_PROPERTY.COLOR && isMeaningfulCSSColor(value)) {
        style[LABEL_STYLE_PROPERTY.COLOR] = cleanCSSValue(value);
    }
};
const applyStyleTextToStyles = (styleText, containerStyle, labelStyle) => {
    if (!styleText) {
        return;
    }
    parseCSSDeclarations(styleText).forEach(({ property, value }) => {
        applyContainerStyleProperty(containerStyle, property, value);
        applyLabelStyleProperty(labelStyle, property, value);
    });
};
const applyStyleTextToLabelStyle = (styleText, labelStyle) => {
    if (!styleText) {
        return;
    }
    parseCSSDeclarations(styleText).forEach(({ property, value }) => {
        if (property === CONTAINER_STYLE_PROPERTY.FILL &&
            isMeaningfulCSSColor(value)) {
            labelStyle[LABEL_STYLE_PROPERTY.COLOR] = cleanCSSValue(value);
            return;
        }
        applyLabelStyleProperty(labelStyle, property, value);
    });
};
const getExplicitStyleProperties = (styleTexts) => {
    const properties = new Set();
    styleTexts.filter(Boolean).forEach((styleText) => {
        parseCSSDeclarations(styleText || "").forEach(({ property }) => {
            properties.add(property);
        });
    });
    return properties;
};
const applyElementAttributesToContainerStyle = (element, containerStyle, explicitProperties) => {
    if (!element) {
        return;
    }
    const attrs = [
        [CONTAINER_STYLE_PROPERTY.FILL, element.getAttribute("fill")],
        [CONTAINER_STYLE_PROPERTY.STROKE, element.getAttribute("stroke")],
        [
            CONTAINER_STYLE_PROPERTY.STROKE_WIDTH,
            element.getAttribute("stroke-width"),
        ],
        [
            CONTAINER_STYLE_PROPERTY.STROKE_DASHARRAY,
            element.getAttribute("stroke-dasharray"),
        ],
    ];
    attrs.forEach(([property, rawValue]) => {
        if (!explicitProperties.has(property)) {
            return;
        }
        if (containerStyle[property]) {
            return;
        }
        const value = cleanCSSValue(rawValue || "");
        if (!value) {
            return;
        }
        applyContainerStyleProperty(containerStyle, property, value);
    });
};
const applyElementAttributesToLabelStyle = (element, labelStyle, explicitProperties) => {
    if (!element) {
        return;
    }
    const candidates = [
        element,
        ...Array.from(element.querySelectorAll("text, foreignObject, div, span, p")),
    ];
    for (const candidate of candidates) {
        if (labelStyle[LABEL_STYLE_PROPERTY.COLOR]) {
            break;
        }
        if (explicitProperties.has(LABEL_STYLE_PROPERTY.COLOR) ||
            explicitProperties.has(CONTAINER_STYLE_PROPERTY.FILL)) {
            applyStyleTextToLabelStyle(candidate.getAttribute("style"), labelStyle);
            if (labelStyle[LABEL_STYLE_PROPERTY.COLOR]) {
                break;
            }
        }
        const color = cleanCSSValue(candidate.getAttribute("fill") || candidate.getAttribute("color") || "");
        if ((explicitProperties.has(LABEL_STYLE_PROPERTY.COLOR) ||
            explicitProperties.has(CONTAINER_STYLE_PROPERTY.FILL)) &&
            isMeaningfulCSSColor(color)) {
            labelStyle[LABEL_STYLE_PROPERTY.COLOR] = color;
        }
    }
};
const accumulateTranslation = (node, stopAt) => {
    let tx = 0;
    let ty = 0;
    let current = node;
    while (current && current !== stopAt) {
        const { transformX, transformY } = getTransformAttr(current);
        tx += transformX;
        ty += transformY;
        current = current.parentElement;
    }
    return { tx, ty };
};
const getAbsoluteBounds = (node, containerEl) => {
    const bbox = node.getBBox();
    const { tx, ty } = accumulateTranslation(node, containerEl);
    return {
        x: bbox.x + tx,
        y: bbox.y + ty,
        width: bbox.width,
        height: bbox.height,
    };
};
const getDividerLine = (nodeEl, containerEl) => {
    const dividerLine = nodeEl.querySelector("line.divider");
    if (!dividerLine) {
        return undefined;
    }
    const { tx, ty } = accumulateTranslation(dividerLine, containerEl);
    return {
        startX: Number(dividerLine.getAttribute("x1")) + tx,
        startY: Number(dividerLine.getAttribute("y1")) + ty,
        endX: Number(dividerLine.getAttribute("x2")) + tx,
        endY: Number(dividerLine.getAttribute("y2")) + ty,
    };
};
const getShapeArea = (element) => {
    const bounds = element.getBBox();
    return Math.abs(bounds.width * bounds.height);
};
const getExplicitStyleValue = (element, property) => {
    const styleText = element.getAttribute("style");
    if (!styleText) {
        return undefined;
    }
    const declaration = parseCSSDeclarations(styleText).find((entry) => entry.property === property);
    if (!declaration) {
        return undefined;
    }
    return cleanCSSValue(declaration.value);
};
const pickShapeElement = (elements, strategy) => {
    const candidates = elements
        .map((element) => ({ element, area: getShapeArea(element) }))
        .filter(({ area }) => Number.isFinite(area) && area > 0);
    if (candidates.length === 0) {
        return null;
    }
    const sortedCandidates = candidates.sort((left, right) => strategy === "largest" ? right.area - left.area : left.area - right.area);
    return sortedCandidates[0].element;
};
const getElementShapeColor = (element, explicitProperties) => {
    if (!element) {
        return undefined;
    }
    if (!explicitProperties.has(CONTAINER_STYLE_PROPERTY.FILL) &&
        !explicitProperties.has(CONTAINER_STYLE_PROPERTY.STROKE)) {
        return undefined;
    }
    const fill = cleanCSSValue(element.getAttribute("fill") ||
        getExplicitStyleValue(element, CONTAINER_STYLE_PROPERTY.FILL) ||
        "");
    const stroke = cleanCSSValue(element.getAttribute("stroke") ||
        getExplicitStyleValue(element, CONTAINER_STYLE_PROPERTY.STROKE) ||
        "");
    if (isMeaningfulCSSColor(fill)) {
        return fill;
    }
    if (isMeaningfulCSSColor(stroke)) {
        return stroke;
    }
    return undefined;
};
const getEndStateInnerColor = (nodeEl, explicitProperties) => {
    const shapeElements = Array.from(nodeEl.querySelectorAll("circle, ellipse, path"));
    const innerShapeElement = pickShapeElement(shapeElements, "smallest");
    return getElementShapeColor(innerShapeElement, explicitProperties);
};
const trimSharedTrailingLineIndentation = (lines) => {
    if (lines.length < 2) {
        return lines;
    }
    const trailingLines = lines.slice(1);
    const commonIndent = trailingLines
        .filter((line) => line.trim().length > 0)
        .reduce((minIndent, line) => {
        const indent = line.match(/^\s*/)?.[0].length ?? 0;
        return Math.min(minIndent, indent);
    }, Number.POSITIVE_INFINITY);
    if (!Number.isFinite(commonIndent) || commonIndent <= 0) {
        return lines.map((line) => line.trimEnd());
    }
    return [
        lines[0].trimEnd(),
        ...trailingLines.map((line) => line.replace(new RegExp(`^\\s{0,${commonIndent}}`), "").trimEnd()),
    ];
};
const getStateLabelText = (node) => {
    const labelLines = Array.isArray(node.label)
        ? node.label.map((entry) => entityCodesToText(entry))
        : entityCodesToText(node.label || "").split("\n");
    const labelText = trimSharedTrailingLineIndentation(labelLines).join("\n");
    return labelText;
};
const getStateDescription = (node) => {
    if (!node.description) {
        return [];
    }
    const description = Array.isArray(node.description)
        ? node.description
        : [node.description];
    return description
        .map((entry) => entityCodesToText(entry))
        .filter((entry) => entry.length > 0);
};
const createNodeElementResolver = (containerEl) => {
    const usedElements = new Set();
    const markAndReturn = (element) => {
        if (element) {
            usedElements.add(element);
        }
        return element;
    };
    const getNextUnusedCandidate = (elements) => {
        const candidate = elements.find((element) => !usedElements.has(element));
        return markAndReturn(candidate || null);
    };
    return (node) => {
        const selectors = [
            `[id='${node.domId}']`,
            `[id='${node.id}']`,
            `[data-id='${node.id}']`,
        ];
        for (const selector of selectors) {
            const element = containerEl.querySelector(selector);
            if (element) {
                return markAndReturn(element);
            }
        }
        // Mermaid generates random ids for anonymous divider sections during parse
        // and regenerates them during render, so exact ids do not always line up.
        switch (node.shape) {
            case "divider":
                return getNextUnusedCandidate(Array.from(containerEl.querySelectorAll("g.statediagram-cluster-alt")));
            case "stateStart":
                return getNextUnusedCandidate(Array.from(containerEl.querySelectorAll("g.node.default")).filter((element) => element.querySelector("circle.state-start")));
            case "stateEnd":
                return getNextUnusedCandidate(Array.from(containerEl.querySelectorAll("g.node.default")).filter((element) => !element.querySelector("circle.state-start")));
            default:
                return null;
        }
    };
};
const findShapeElement = (nodeEl, shape) => {
    switch (shape) {
        case "roundedWithTitle":
            return (nodeEl.querySelector("rect.outer") ||
                nodeEl.querySelector("rect") ||
                nodeEl);
        case "divider":
            return (nodeEl.querySelector("rect.divider") ||
                nodeEl.querySelector("rect") ||
                nodeEl);
        case "rectWithTitle":
            return (nodeEl.querySelector("rect.outer") ||
                nodeEl.querySelector("rect") ||
                nodeEl);
        case "stateStart":
            return (pickShapeElement(Array.from(nodeEl.querySelectorAll("circle, ellipse, path")), "largest") || nodeEl);
        case "stateEnd":
            return (pickShapeElement(Array.from(nodeEl.querySelectorAll("circle, ellipse, path")), "largest") || nodeEl);
        case "choice":
        case "fork":
        case "join":
        case "note":
        case "rect":
        default:
            return (nodeEl.querySelector("rect, path, circle, ellipse, polygon") || nodeEl);
    }
};
const parseStateNode = (node, containerEl, resolveNodeElement) => {
    const nodeEl = resolveNodeElement(node);
    if (!nodeEl) {
        throw new Error(`State node element not found for "${node.id}"`);
    }
    const shapeEl = findShapeElement(nodeEl, node.shape);
    const containerStyle = {};
    const labelStyle = {};
    const explicitStyleTexts = [
        node.labelStyle,
        ...(node.cssCompiledStyles || []),
        ...(node.cssStyles || []),
    ];
    const explicitStyleProperties = getExplicitStyleProperties(explicitStyleTexts);
    explicitStyleTexts.filter(Boolean).forEach((styleText) => {
        applyStyleTextToStyles(styleText, containerStyle, labelStyle);
    });
    applyElementAttributesToContainerStyle(shapeEl, containerStyle, explicitStyleProperties);
    applyElementAttributesToLabelStyle(nodeEl, labelStyle, explicitStyleProperties);
    const bounds = getAbsoluteBounds(shapeEl, containerEl);
    return {
        id: node.id,
        shape: node.shape,
        text: getStateLabelText(node),
        description: getStateDescription(node),
        x: bounds.x,
        y: bounds.y,
        width: bounds.width,
        height: bounds.height,
        parentId: node.parentId,
        position: node.position,
        containerStyle,
        labelStyle,
        dividerLine: node.shape === "rectWithTitle"
            ? getDividerLine(nodeEl, containerEl)
            : undefined,
        endInnerColor: node.shape === "stateEnd"
            ? getEndStateInnerColor(nodeEl, explicitStyleProperties)
            : undefined,
        isRenderable: node.shape !== "noteGroup",
    };
};
const parseStateEdge = (edge, containerEl) => {
    const edgeEl = containerEl.querySelector(`[id='${edge.id}']`);
    if (!edgeEl) {
        return null;
    }
    const { tx, ty } = accumulateTranslation(edgeEl, containerEl);
    const edgePositionData = computeEdgePositions(edgeEl, { x: tx, y: ty }, "MCL");
    if (edgePositionData.reflectionPoints.length < 2) {
        return null;
    }
    const edgeStyle = {};
    const applyEdgeStyleProperty = (property, value) => {
        switch (property) {
            case CONTAINER_STYLE_PROPERTY.STROKE:
                if (isMeaningfulCSSColor(value)) {
                    edgeStyle.strokeColor = cleanCSSValue(value);
                }
                break;
            case CONTAINER_STYLE_PROPERTY.STROKE_WIDTH: {
                const strokeWidth = parseFloat(cleanCSSValue(value));
                if (Number.isFinite(strokeWidth) && strokeWidth > 0) {
                    edgeStyle.strokeWidth = strokeWidth;
                }
                break;
            }
            case CONTAINER_STYLE_PROPERTY.STROKE_DASHARRAY:
                if (cleanCSSValue(value)) {
                    edgeStyle.strokeStyle = "dashed";
                }
                break;
        }
    };
    [edge.style].filter(Boolean).forEach((styleText) => {
        parseCSSDeclarations(styleText || "").forEach(({ property, value }) => {
            applyEdgeStyleProperty(property, value);
        });
    });
    const isNoteEdge = edge.arrowhead === "none" || edge.classes?.includes("note-edge");
    return {
        id: edge.id,
        start: edge.start,
        end: edge.end,
        text: entityCodesToText(edge.label || ""),
        ...edgePositionData,
        strokeColor: edgeStyle.strokeColor,
        strokeWidth: edgeStyle.strokeWidth,
        strokeStyle: isNoteEdge ? "dashed" : edgeStyle.strokeStyle,
        isNoteEdge,
    };
};
export const parseMermaidStateDiagram = (db, containerEl) => {
    const { nodes, edges } = db.getData();
    const resolveNodeElement = createNodeElementResolver(containerEl);
    return {
        type: "state",
        nodes: nodes.map((node) => parseStateNode(node, containerEl, resolveNodeElement)),
        edges: edges
            .map((edge) => parseStateEdge(edge, containerEl))
            .filter((edge) => edge !== null),
    };
};
