// mermaid/dist/chunks/mermaid.core/classDiagram-OUVF2IWQ.mjs
import {
  ClassDB,
  classDiagram_default,
  classRenderer_v3_unified_default,
  styles_default
} from "./chunk-V7JOEXUC.mjs";
import "./chunk-5VM5RSS4.mjs";
import "./chunk-XXDRQBXY.mjs";
import "./chunk-VR4S4FIN.mjs";
import "./chunk-FWX5IMBZ.mjs";
import "./chunk-32BRIVSS.mjs";
import "./chunk-52WLFC77.mjs";
import "./chunk-ZGVPDNZ5.mjs";
import "./chunk-C7G6YPKG.mjs";
import "./chunk-7BUUIJ7U.mjs";
import "./chunk-OGEWGWER.mjs";
import "./chunk-Q4XR5HBZ.mjs";
import "./chunk-HOUHSVGY.mjs";
import "./chunk-ICXQ74PX.mjs";
import "./chunk-WYO6CB5R.mjs";
import "./chunk-X3CZISLH.mjs";
import {
  __name
} from "./chunk-Y2CYZVJY.mjs";

// src/diagrams/class/classDiagram.ts
var diagram = {
  parser: classDiagram_default,
  get db() {
    return new ClassDB();
  },
  renderer: classRenderer_v3_unified_default,
  styles: styles_default,
  init: /* @__PURE__ */ __name((cnf) => {
    if (!cnf.class) {
      cnf.class = {};
    }
    cnf.class.arrowMarkerAbsolute = cnf.arrowMarkerAbsolute;
  }, "init")
};
export {
  diagram
};
