// mermaid/dist/chunks/mermaid.core/stateDiagram-v2-6OUMAXLB.mjs
import {
  StateDB,
  stateDiagram_default,
  stateRenderer_v3_unified_default,
  styles_default
} from "./chunk-EX3LRPZG.mjs";
import "./chunk-XXDRQBXY.mjs";
import "./chunk-VR4S4FIN.mjs";
import "./chunk-FWX5IMBZ.mjs";
import "./chunk-32BRIVSS.mjs";
import "./chunk-52WLFC77.mjs";
import "./chunk-ZGVPDNZ5.mjs";
import "./chunk-C7G6YPKG.mjs";
import "./chunk-7BUUIJ7U.mjs";
import "./chunk-OGEWGWER.mjs";
import "./chunk-Q4XR5HBZ.mjs";
import "./chunk-HOUHSVGY.mjs";
import "./chunk-ICXQ74PX.mjs";
import "./chunk-WYO6CB5R.mjs";
import "./chunk-X3CZISLH.mjs";
import {
  __name
} from "./chunk-Y2CYZVJY.mjs";

// src/diagrams/state/stateDiagram-v2.ts
var diagram = {
  parser: stateDiagram_default,
  get db() {
    return new StateDB(2);
  },
  renderer: stateRenderer_v3_unified_default,
  styles: styles_default,
  init: /* @__PURE__ */ __name((cnf) => {
    if (!cnf.state) {
      cnf.state = {};
    }
    cnf.state.arrowMarkerAbsolute = cnf.arrowMarkerAbsolute;
  }, "init")
};
export {
  diagram
};
