// mermaid/dist/chunks/mermaid.esm/stateDiagram-v2-66XI6PKQ.mjs
import {
  StateDB,
  stateDiagram_default,
  stateRenderer_v3_unified_default,
  styles_default
} from "./chunk-5J66UF2J.mjs";
import "./chunk-HPLX5OYV.mjs";
import "./chunk-DRARJEGU.mjs";
import "./chunk-FW7PRHBN.mjs";
import "./chunk-QVUG6NDL.mjs";
import "./chunk-4F4KDU6L.mjs";
import "./chunk-65BZPYT2.mjs";
import "./chunk-PLCLPJVV.mjs";
import "./chunk-BNCO5QFQ.mjs";
import "./chunk-RTI7CJYH.mjs";
import "./chunk-3UJ2IBUM.mjs";
import "./chunk-JCJHR2HC.mjs";
import "./chunk-4PNYCQBS.mjs";
import "./chunk-MMGVDTGO.mjs";
import "./chunk-IPM4HZQ6.mjs";
import "./chunk-CHAKFXHA.mjs";
import "./chunk-FO5PYUIK.mjs";
import {
  __name
} from "./chunk-PTVI3W5X.mjs";

// src/diagrams/state/stateDiagram-v2.ts
var diagram = {
  parser: stateDiagram_default,
  get db() {
    return new StateDB(2);
  },
  renderer: stateRenderer_v3_unified_default,
  styles: styles_default,
  init: /* @__PURE__ */ __name((cnf) => {
    if (!cnf.state) {
      cnf.state = {};
    }
    cnf.state.arrowMarkerAbsolute = cnf.arrowMarkerAbsolute;
  }, "init")
};
export {
  diagram
};
